### PR TITLE
enhance exception handling

### PR DESCRIPTION
**Pull request by mamin506**
1. Add cacheline aligned submitted and completed counter in KDS. (They are counting on critical path, but I had verified that is has no performance impact)
2. KDS core logic would use above counters to detect if there are outstanding commands running in CU/ERT sub-devices.
3. Add periodic timer per CU sub-dev. It is used to determine if a command is timeout.

### CODE REVIEW

**Review by chienwei-lan — APPROVED** [review]

File: src/runtime_src/core/common/drv/xrt_cu.c [hunk 1]
```diff
@@ -90,13 +117,45 @@ static inline void __process_sq(struct xrt_cu *xcu)
...
+		} else if (unlikely(ev_client)) {
+			/* Client event happens rarely */
+			if (xcmd->client != ev_client)
+				continue;
+
+			tick = atomic_read(&xcu->tick);
+			/* Record CU tick to start timeout counting */
+			if (!xcmd->tick) {
+				xcmd->tick = tick;
+				xcmd->ttl = CU_SEC2TIMER(CU_EXEC_DEFAULT_TTL);
```
Comment (maxzhen) [inline src/runtime_src/core/common/drv/xrt_cu.c:143]: do you really need xcmd->ttl? is it just a hard-coded value anyway and never change?

File: src/runtime_src/core/common/drv/kds_core.c [hunk 2]
```diff
@@ -371,29 +373,36 @@ kds_del_cu_context(struct kds_sched *kds, struct kds_client *client,
...
+	if (!kds->ert_disable) {
+		kds->ert->abort_done(kds->ert, client, cu_idx);
+		bad_state = kds->ert->is_bad_state(kds->ert);
+	} else {
+		xrt_cu_abort_done(cu_mgmt->xcus[cu_idx], client);
+		bad_state = xrt_cu_is_bad_state(cu_mgmt->xcus[cu_idx]);
+	}
+
+	if (bad_state) {
 		kds->bad_state = 1;
```
Comment (maxzhen) [inline src/runtime_src/core/common/drv/kds_core.c:404]: where do we start aborting all cmds on this device once the state becomes bad?

File: src/runtime_src/core/common/drv/kds_core.c [hunk 3]
```diff
@@ -371,29 +373,36 @@ kds_del_cu_context(struct kds_sched *kds, struct kds_client *client,
...
 
-	if (state == CU_STATE_BAD) {
-		kds_info(client, "CU(%d) hangs, please reset device", cu_idx);
-		/* No lock to protect bad_state.
-		 * If a new client doesn't get the updated status and send commands,
-		 * those commands would failed with TIMEOUT state.
-		 */
+	if (!kds->ert_disable) {
+		kds->ert->abort_done(kds->ert, client, cu_idx);
+		bad_state = kds->ert->is_bad_state(kds->ert);
```
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/common/drv/kds_core.c:397]: Ok, sure.

File: src/runtime_src/core/common/drv/kds_core.c (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/core/common/drv/kds_core.c:404]: In xocl_kds.c, xocl_command_ioctl(). If state becomes bad, the execbuf ioctl would directly return -EDEADLK without allocating xcmd.

File: src/runtime_src/core/common/drv/xrt_cu.c (hunk 1 above)
Comment (mamin506) [reply src/runtime_src/core/common/drv/xrt_cu.c:143]: I don't really need that for now. I would just use a hard-code value before command timeout is configurable.

File: src/runtime_src/core/common/drv/kds_core.c [hunk 4]
```diff
@@ -371,29 +373,36 @@ kds_del_cu_context(struct kds_sched *kds, struct kds_client *client,
 		return -EINVAL;
 	}
 
-	/* TODO: finish ERT abort process later */
-	if (!kds->ert_disable)
+	/* Before close, make sure no remain commands in CU's queue. */
+	outstanding = client->s_cnt[cu_idx] - client->c_cnt[cu_idx];
```
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/common/drv/kds_core.c:387]: This is a good point. After reconsidering it, I fix a potential race condition.

kds_del_cu_context(),  would do test_and_clear_bit(cu_idx, client->cu_bitmap) before we get here.
In acquire_cu_idx(), s_cnt would increase after test_bit(cu_idx, client->cu_bitmap) check.
Then acquire_cu_idx() would do test_bit(cu_idx, client->cu_bitmap) again. This make sure only below sequence can get execute.

acquire_cu_idx()                         kds_del_cu_context()
          .                                                      .
     s_cnt++                                               .
          .                                                      .
    test_bit()                                               .
          .                                          test_and_clear_bit()

**Comment by mamin506** [comment]
I have test this change with slow CU with KDS mode. It looks good.

**Review by maxzhen — APPROVED** [review]

File: src/runtime_src/core/common/drv/xrt_cu.c [hunk 5]
```diff
@@ -329,11 +350,10 @@ int xrt_cu_intr_thread(void *data)
 			continue;
 
 		if (xcu->num_sq || is_zero_credit(xcu)) {
-			//usleep_range(1, 3);
 			xrt_cu_check(xcu);
 			if (!xcu->done_cnt || !xcu->ready_cnt) {
 				xcu->sleep_cnt++;
-				if (down_interruptible(&xcu->sem_cu))
+				if (down_timeout(&xcu->sem_cu, CU_TIMER))
```
Comment (maxzhen) [inline src/runtime_src/core/common/drv/xrt_cu.c:356]: I missed this change. Why do you need to change down_interruptible() to down_timeout()?

Comment (xilinxgitops) [reply src/runtime_src/core/common/drv/xrt_cu.c:356]: I am not retrying this PR. Kindly post in XRT channel if you want this PR to be retried - Palash

Comment (mamin506) [reply src/runtime_src/core/common/drv/xrt_cu.c:356]: If a kernel runs for a long time or hang, ctrl-c cannot interrupt CU thread so this CU's c_cnt has no chance to increase.

**Pull request merged**